assets/backport/github/github.go

```diff
@@ -0,0 +1,311 @@
+		c.Organization,
+		c.Repository,
+		branch.GetCommit().GetSHA())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	for i := 0; i < len(commits); i++ {
+		cherryCommit, _, err := c.Client.Git.GetCommit(ctx, c.Organization, c.Repository, commits[i])
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		// Skip merge commits. The commit to cherry-pick MUST have only 1 parent.
+		if len(cherryCommit.Parents) != 1 {
+			continue
+		}
+		tree, sha, err := c.cherryPickCommit(ctx, branchName, cherryCommit, headCommit)
+		if err != nil {
+			return trace.Errorf("please manually delete branch %s: %v", branchName, err)
+		}
+		headCommit.SHA = &sha
+		headCommit.Tree = tree
+	}
+	return nil
+}
+
+// cherryPickCommit cherry picks a single commit on a branch.
+func (c *Client) cherryPickCommit(ctx context.Context, branchName string, cherryCommit, headBranchCommit *go_github.Commit) (*go_github.Tree, string, error) {
+	cherryParent := cherryCommit.Parents[0]
+
+	// Temporarily set the parent of the branch HEAD to the parent of the commit
+	// to cherry-pick so they are siblings.
+	err := c.createSiblingCommit(ctx, branchName, headBranchCommit, cherryParent)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+
+	// When git performs the merge, it detects that the parent of the branch commit that is
+	// being merged onto matches the parent of the cherry pick commit, and merges a tree of size 1.
+	// The merge commit will contain the delta between the file tree in target branch and the
+	// commit to cherry-pick.
+	merge, err := c.merge(ctx, branchName, cherryCommit.GetSHA())
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+	mergeTree := merge.GetTree()
+
+	updatedCommit, _, err := c.Client.Git.GetCommit(ctx,
+		c.Organization,
+		c.Repository,
+		headBranchCommit.GetSHA())
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+	// Create the actual cherry-pick commit on the target branch containing the merge commit tree.
+	commit, _, err := c.Client.Git.CreateCommit(ctx, c.Organization, c.Repository, &go_github.Commit{
+		Message: cherryCommit.Message,
+		Tree:    mergeTree,
+		Parents: []*go_github.Commit{
+			updatedCommit,
+		},
+	})
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+
+	// Overwrite the merge commit and its parent on the branch by the newly created commit.
+	// The result will be equivalent to what would have happened with a fast-forward merge.
+	sha := commit.GetSHA()
+	refName := fmt.Sprintf("%s%s", branchRefPrefix, branchName)
+	_, _, err = c.Client.Git.UpdateRef(ctx, c.Organization, c.Repository, &go_github.Reference{
+		Ref: go_github.String(refName),
+		Object: &go_github.GitObject{
+			SHA: go_github.String(sha),
+		},
+	}, true)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+	return mergeTree, sha, nil
+}
+
+// createSiblingCommit creates a commit with the passed in commit's tree and parent
+// and updates the passed in branch to point at that commit.
```

Comment:
I think some ASCII art might help here, or in the body of the function at least.

Reply:
What are some of your ideas? I feel ASCII art will add a lot of clutter to the code.